### PR TITLE
Again my repo is bugged out, ignore the insecent "x commits ahead"

### PR DESCRIPTION
**Pull request by sokorototo**
- So I made a very minor change to `vach`, removing a mutable reference where it wasn't used.
 - Most changes are however to `vach-cli`. I refactored the code, so now we have `app`, `command`, `utils` and `keys` submodules. Instead of all the logic being in `config.rs`. Which has been separated. Also, since the subcommands are trait bound, it is now easier to add a new subcommand. Overall the code is still your original code, only heavily refactored and made spec compliant.

### CODE REVIEW

File: vach-cli/Cargo.toml [hunk 1]
```diff
@@ -39,4 +36,6 @@ lazy_static = "1.4.0"
 tabled = "0.3.0"
 log = "0.4.14"
 pretty_env_logger = "0.4.0"
-vach = "0.2.2"
+vach = {	path = "../vach"	}
```
Comment (zeskeertwee) [inline vach-cli/Cargo.toml:39]: we need to use a version availible on crates.io for it to be published

File: vach/src/writer/builder/config.rs [hunk 2]
```diff
@@ -1,17 +1,48 @@
...
 	/// Used to write a unique magic sequence into the write target.
 	pub magic: [u8; crate::MAGIC_LENGTH],
 	/// Flags to be written into the `Header` section of the write target.
 	pub flags: Flags,
 	/// An optional keypair. If a key is provided, then the write target will have signatures for tamper verification.
 	pub keypair: Option<esdalek::Keypair>,
+	/// An optional callback that is called every time a `Leaf` finishes processing
+	/// The type signature is:
+	/// ```ignore
+	/// type OptionalCallback = Option<Box<dyn Fn(&String, usize, &RegistryEntry)>>;
```
Comment (zeskeertwee) [inline vach/src/writer/builder/config.rs:20]: we should probably annotate what every parameter means in the callback

File: vach-cli/Cargo.toml [hunk 3]
```diff
@@ -1,8 +1,8 @@
 [package]
 name = "vach-cli"
-version = "0.2.2"
+version = "0.3.0"
```
Comment (zeskeertwee) [inline vach-cli/Cargo.toml:3]: it might be a good idea to keep this version running alongside with the vach version, or is that a bad idea?

**Review by zeskeertwee — CHANGES_REQUESTED** [review]

File: vach-cli/Cargo.toml (hunk 1 above)
Comment (sokorototo) [reply vach-cli/Cargo.toml:39]: I had to modify vach and add some utility functions to vach, the CLI depends on those features. Once I publish the latest version of vach. then I will switch to a pathless dependency

File: vach/src/writer/builder/config.rs (hunk 2 above)
Comment (sokorototo) [reply vach/src/writer/builder/config.rs:20]: And how do you do that?

File: vach-cli/Cargo.toml (hunk 3 above)
Comment (sokorototo) [reply vach-cli/Cargo.toml:3]: Version 0.3.0 is the first spec-compliant version, it marks a new leaf really. Plus vach changed a bit

File: vach/src/writer/builder/config.rs (hunk 2 above)
Comment (zeskeertwee) [reply vach/src/writer/builder/config.rs:20]: oh, i just meant writing it in the documentation

Comment (sokorototo) [reply vach/src/writer/builder/config.rs:20]: Oh ok

Comment (sokorototo) [reply vach/src/writer/builder/config.rs:20]: I also published vach 0.2.3. The changes are insignificant

Comment (zeskeertwee) [reply vach/src/writer/builder/config.rs:20]: Okay, could you add a commit to this branch that adds some documentation?

Comment (sokorototo) [reply vach/src/writer/builder/config.rs:20]: Leave that to the next commit. This is technically the last commit from my fork before I replace it with a working fork

**Review by zeskeertwee — APPROVED** [review]

**Comment by sokorototo** [comment]
Saying ggodbye to sokorototo-main

**Pull request merged**